### PR TITLE
Fix #327: add upgrade manager tests

### LINKED ISSUE #327: upgrades.rb:12-15: Write the upgrade manager tests that...
The puzzle `285-214c18b0` from #285 has to be resolved:

https://github.com/zold-io/zold/blob/90f141e718477fbbfcf49542726ad7204f70fcf7/lib/zold/upgrades.rb#L12-L15

The puzzle was created by James C Russell on 30-Jun-18. 

Estimate: 30 minutes,     

If you have any technical questions, don't ask me, submit new tickets instead. The task will be "done" when the problem is fixed and the text of the puzzle is _removed_ from the source code. Here is more about [PDD](http://www.yegor256.com/2009/03/04/pdd.html) and [about me](http://www.yegor256.com/2017/04/05/pdd-in-action.html).

### PR DESCRIPTION
**Pull request by ledestin**
https://github.com/zold-io/zold/issues/327

I'm having real upgrade scripts run and output `upgrading to <version>`, which I'm checking in tests.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#351](https://github.com/zold-io/zold/issues/351) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#351](https://github.com/zold-io/zold/issues/351) is assigned to @botanicus[/z](https://www.0crat.com/u/botanicus), here is [why](https://www.0crat.com/footprint/CAZPZR9FS/1861019805); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by botanicus** [comment]
@ledestin I might be wrong about this one, but if I remember correctly @yegor256 said something about not using `”Fix #id”` in commit messages. But I might be wrong.

File: test/test_upgrades.rb [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+require 'minitest/autorun'
+require_relative 'test__helper'
+require_relative '../lib/zold/upgrades'
+require_relative '../lib/zold/version_file'
+
+# Upgrade test.
+# Author:: Yegor Bugayenko (yegor256@gmail.com)
+# Copyright:: Copyright (c) 2018 Yegor Bugayenko
+# License:: MIT
+class TestUpgrade < Minitest::Test
```
Comment (botanicus) [inline test/test_upgrades.rb]: @ledestin If the file is called `test_upgrades.rb`, it should probably be called `TestUpgrades` for consistency.

**Comment by botanicus** [comment]
@ledestin thanks, looks good!

@yegor256 it’s good to be merged.

File: test/test_upgrades.rb [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  def version_file(dir)
+    Zold::VersionFile.new(File.join(dir, 'version'))
+  end
+
+  def create_version_file(dir, version)
+    IO.write(File.join(dir, 'version'), version)
+  end
+
+  def create_upgrade_file(dir, version)
+    upgrade_file_name = File.join(dir, "#{version}.rb")
```
Comment (yegor256) [inline test/test_upgrades.rb]: @ledestin see https://www.yegor256.com/2015/09/01/redundant-variables-are-evil.html

File: test/test_upgrades.rb [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+require_relative '../lib/zold/upgrades'
+require_relative '../lib/zold/version_file'
+
+# Upgrade test.
+# Author:: Yegor Bugayenko (yegor256@gmail.com)
+# Copyright:: Copyright (c) 2018 Yegor Bugayenko
+# License:: MIT
+class TestUpgrades < Minitest::Test
+  def test_no_version_file_is_ok
+    skip
```
Comment (yegor256) [inline test/test_upgrades.rb]: @ledestin I would expect a `todo` here, otherwise what is the point of this test?

File: test/test_upgrades.rb [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      create_upgrade_file(dir, script_version)
+      expected_output = expected_upgrade_script_output(script_version)
+      assert_output(/#{expected_output}/) do
+        run_upgrades(dir)
+      end
+    end
+  end
+
+  def test_pending_scripts_run
+    skip
```
Comment (yegor256) [inline test/test_upgrades.rb]: @ledestin I would expect a `todo` here, otherwise what is the point of this test?

File: test/test_upgrades.rb [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    end
+  end
+
+  def test_already_ran_scripts_dont_run
+    Dir.mktmpdir do |dir|
+      %w[0.0.1 0.0.2].each do |script_version|
+        create_upgrade_file(dir, script_version)
+      end
+      create_version_file(dir, '0.0.1')
+      must_not_be_in_output = expected_upgrade_script_output('0.0.1')
```
Comment (yegor256) [inline test/test_upgrades.rb]: @ledestin check this out: https://www.yegor256.com/2015/01/12/compound-name-is-code-smell.html

**Comment by yegor256** [comment]
@ledestin @botanicus see above. The biggest problem here is that we are introducing two new tests which are skipped and we don't have any "puzzles". They will never be unskipped, since nobody will be interested to do that. We have to create puzzles and explain what these tests are for and what should be done in order to make them work.

Second, please pay attention to the links I posted. Long variable names is a code smell. Redundant variables is a code smell.

**Comment by botanicus** [comment]
@yegor256 thanks, understood.

File: test/test_upgrades.rb (hunk 5 above)
Comment (ledestin) [reply test/test_upgrades.rb]: @yegor256 what would you call it? `output` would be bad, in my opinion.

**Comment by ledestin** [comment]
@yegor256 thanks, please see updates. Regarding `script_version`, I'd have a name clash with version file if I used just one word.

**Comment by yegor256** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @botanicus[/z](https://www.0crat.com/u/botanicus), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#351](https://github.com/zold-io/zold/issues/351) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256)

**Comment by ypshenychka** [comment]
@botanicus According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>The code reviewer found at least three problems in the code.
Comments were mostly about design problems, not cosmetic issues.

Only one issue was found during code review. 
Please confirm that you'll try to find at least three major problems while future reviews.

**Comment by ypshenychka** [comment]
@botanicus thanks

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +15 point(s) just awarded to @botanicus[/z](https://www.0crat.com/u/botanicus)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Comment by ledestin** [comment]
@yegor256 I forgot to remove the puzzle, what should I do?

**Comment by yegor256** [comment]
@ledestin submit a new pull request

**Pull request merged**